In [2]:
train_path = r'dataset_Sep_1_2025\train.csv'
test_path = r'dataset_Sep_1_2025\test.csv'
old_repos_path = r'dataset_Sep_1_2025\DeepFunding Repos Enhanced via OpenQ - EXCLUDED _ NOT COMMIT AFTER 2020.csv'
big_org_path = r'dataset_Sep_1_2025\DeepFunding Repos Enhanced via OpenQ - EXCLUDED_ TOO LARGE _ BIG ORGS.csv'
enhanced_teams_path = r'dataset_Sep_1_2025\DeepFunding Repos Enhanced via OpenQ - ENHANCED TEAMS.csv'
enhanced_contributors_path = r'dataset_Sep_1_2025\DeepFunding Repos Enhanced via OpenQ - ENHANCED CONTRIBUTORS.csv'

In [3]:
import pandas as pd
train = pd.read_csv(train_path)
test = pd.read_csv(test_path)
old_repos = pd.read_csv(old_repos_path)
big_org = pd.read_csv(big_org_path)
enhanced_teams = pd.read_csv(enhanced_teams_path)
enhanced_contributors = pd.read_csv(enhanced_contributors_path)

In [4]:
contributor_names = enhanced_teams.assign(Names = enhanced_teams['recentContributors'].str.split(', ')).explode('Names')

contributor_names

owner           name dependencies  \
0                    oven-sh            bun    no status   
0                    oven-sh            bun    no status   
0                    oven-sh            bun    no status   
0                    oven-sh            bun    no status   
0                    oven-sh            bun    no status   
...                      ...            ...          ...   
4422                  coreos     go-systemd    no status   
4423             cockroachdb         errors    no status   
4424              manifoldco       promptui    no status   
4425             cockroachdb        logtags    no status   
4426  census-instrumentation  opencensus-go    no status   

                                            description latestCommit  \
0     Incredibly fast JavaScript runtime, bundler, t...     3/1/2025   
0     Incredibly fast JavaScript runtime, bundler, t...     3/1/2025   
0     Incredibly fast JavaScript runtime, bundler, t...     3/1/2025   
0     Incredibly fast JavaScript runtime, bundler, t...     3/1/2025   
0     Incredibly fast JavaScript runtime, bundler, t...     3/1/2025   
...                                                 ...          ...   
4422  Go bindings to systemd socket activation, jour...   no commits   
4423  Go error library with error portability over t...     6/3/2024   
4424   Interactive prompt for command-line applications   no commits   
4425              key/value annotations for Go contexts   12/15/2024   
4426  A stats collection and distributed tracing fra...   no commits   

                                             githubLink  \
0                        https://github.com/oven-sh/bun   
0                        https://github.com/oven-sh/bun   
0                        https://github.com/oven-sh/bun   
0                        https://github.com/oven-sh/bun   
0                        https://github.com/oven-sh/bun   
...                                                 ...   
4422               https://github.com/coreos/go-systemd   
4423              https://github.com/cockroachdb/errors   
4424             https://github.com/manifoldco/promptui   
4425             https://github.com/cockroachdb/logtags   
4426  https://github.com/census-instrumentation/open...   

                  websiteLink  reputation activity  popularity  \
0              https://bun.sh          10      75%          10   
0              https://bun.sh          10      75%          10   
0              https://bun.sh          10      75%          10   
0              https://bun.sh          10      75%          10   
0              https://bun.sh          10      75%          10   
...                       ...         ...      ...         ...   
4422                      NaN           6       0%          10   
4423                      NaN           8       0%          10   
4424  https://www.manifold.co           5       0%          10   
4425                      NaN           8       0%           4   
4426     http://opencensus.io           6       0%          10   

                                       languages  commitCount  totalStars  \
0     Zig, JavaScript, TypeScript, CSS, Makefile        352.0     76508.0   
0     Zig, JavaScript, TypeScript, CSS, Makefile        352.0     76508.0   
0     Zig, JavaScript, TypeScript, CSS, Makefile        352.0     76508.0   
0     Zig, JavaScript, TypeScript, CSS, Makefile        352.0     76508.0   
0     Zig, JavaScript, TypeScript, CSS, Makefile        352.0     76508.0   
...                                          ...          ...         ...   
4422                                   Go, Shell          0.0      2474.0   
4423                                          Go          3.0      2168.0   
4424                                Go, Makefile          0.0      6166.0   
4425                                          Go          2.0        39.0   
4426                   Go, Shell, HTML, Makefile          0.0      2049.0   

          

In [5]:
contributor_names['Names'].isna().sum()

np.int64(1794)

In [6]:
enhanced_contributors

name dependencies                      email  \
0           maxam2017    no status          whk0313@gmail.com   
1              7silva    no status    zaronymogamer@gmail.com   
2       rogerhardiman    no status      opensource@rjh.org.uk   
3            riskrole    no status         yuhang@before.tech   
4        srikrsna-buf    no status         skrishna@buf.build   
...               ...          ...                        ...   
12967       cuiweixie    no status           523516579@qq.com   
12968          saolyn    no status            sammy@rosso.red   
12969           gsora    no status               me@gsora.xyz   
12970  prestonvanloon    no status  preston@prysmaticlabs.com   
12971   samuel-videau    no status    samuel.videau@efrei.net   

                                             description        location  \
0                                   Frontend Developer 🚀          Taiwan   
1      🚀 Full-Stack Developer | Habilidades em Node.J...          Brazil   
2                                      Software Engineer              UK   
3                                                    NaN             NaN   
4                                                    NaN             NaN   
...                                                  ...             ...   
12967                                                NaN  Shenzhen,China   
12968                                                NaN     Switzerland   
12969           Software, bikes and various other stuff.           Italy   
12970  Working on @ethereum @prysmaticlabs @OffchainL...   Nashville, TN   
12971                                                NaN             NaN   

         organization latestCommit                          twitterLink  \
0                 NaN     2/2/2025                                  NaN   
1                 NaN    4/20/2024                                  NaN   
2                 NaN    9/10/2024                                  NaN   
3                 NaN    9/26/2024                                  NaN   
4          connectrpc    2/26/2025                                  NaN   
...               ...          ...                                  ...   
12967  riddles-are-us    1/29/2025                                  NaN   
12968   prysmaticlabs    2/19/2025                                  NaN   
12969             NaN     1/7/2025                                  NaN   
12970   prysmaticlabs    2/24/2025  https://twitter.com/preston_vanloon   
12971             NaN   10/12/2024                                  NaN   

                              githubLink           websiteLink  reputation  \
0           https://github.com/maxam2017     https://maxam.dev           1   
1              https://github.com/7silva                   NaN           3   
2       https://github.com/rogerhardiman        www.rjh.org.uk           4   
3            https://github.com/riskrole                   NaN           0   
4        https://github.com/srikrsna-buf                   NaN           6   
...                                  ...                   ...         ...   
12967       https://github.com/cuiweixie                   NaN           7   
12968          https://github.com/saolyn                   NaN           8   
12969           https://github.com/gsora  http://noskills.club           2   
12970  https://github.com/prestonvanloon       https://pvl.dev           8   
12971   https://github.com/samuel-videau                   NaN           0   

      activity depActivity                            languages  commitCount  \
0          77%          0%                    Shell, TypeScript         49.0   
1          52%          0%         Go, Vue, JavaScript, Unknown         10.0   
2          38%          0%                  C#, JavaScript, C++          4.0   
3          50%          0%  TypeScript, JavaScript, Unknown, Go          9.0   
4          78%          0%           TypeScript, Starlark, Dart         54.0   
...  

In [7]:
repo_with_contributor = pd.merge(
    contributor_names,
    enhanced_contributors,
    left_on = 'Names',
    right_on = 'name',
    how = 'left'
)

repo_with_contributor

owner         name_x dependencies_x  \
0                     oven-sh            bun      no status   
1                     oven-sh            bun      no status   
2                     oven-sh            bun      no status   
3                     oven-sh            bun      no status   
4                     oven-sh            bun      no status   
...                       ...            ...            ...   
20880                  coreos     go-systemd      no status   
20881             cockroachdb         errors      no status   
20882              manifoldco       promptui      no status   
20883             cockroachdb        logtags      no status   
20884  census-instrumentation  opencensus-go      no status   

                                           description_x latestCommit_x  \
0      Incredibly fast JavaScript runtime, bundler, t...       3/1/2025   
1      Incredibly fast JavaScript runtime, bundler, t...       3/1/2025   
2      Incredibly fast JavaScript runtime, bundler, t...       3/1/2025   
3      Incredibly fast JavaScript runtime, bundler, t...       3/1/2025   
4      Incredibly fast JavaScript runtime, bundler, t...       3/1/2025   
...                                                  ...            ...   
20880  Go bindings to systemd socket activation, jour...     no commits   
20881  Go error library with error portability over t...       6/3/2024   
20882   Interactive prompt for command-line applications     no commits   
20883              key/value annotations for Go contexts     12/15/2024   
20884  A stats collection and distributed tracing fra...     no commits   

                                            githubLink_x  \
0                         https://github.com/oven-sh/bun   
1                         https://github.com/oven-sh/bun   
2                         https://github.com/oven-sh/bun   
3                         https://github.com/oven-sh/bun   
4                         https://github.com/oven-sh/bun   
...                                                  ...   
20880               https://github.com/coreos/go-systemd   
20881              https://github.com/cockroachdb/errors   
20882             https://github.com/manifoldco/promptui   
20883             https://github.com/cockroachdb/logtags   
20884  https://github.com/census-instrumentation/open...   

                 websiteLink_x  reputation_x activity_x  popularity  ...  \
0               https://bun.sh            10        75%          10  ...   
1               https://bun.sh            10        75%          10  ...   
2               https://bun.sh            10        75%          10  ...   
3               https://bun.sh            10        75%          10  ...   
4               https://bun.sh            10        75%          10  ...   
...                        ...           ...        ...         ...  ...   
20880                      NaN             6         0%          10  ...   
20881                      NaN             8         0%          10  ...   
20882  https://www.manifold.co             5         0%          10  ...   
20883                      NaN             8         0%           4  ...   
20884     http://opencensus.io             6         0%          10  ...   

                               twitterLink  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3      https://twitter.com/BjornProgrammer   
4                                      NaN   
...                                    ...   
20880                                  NaN   
20881                                  NaN   
20882                                  NaN   
20883                                  NaN   
20884                                  NaN   

                                githubLink_y  \
0               https://github.com/jprinaldi   
1                https://github.com/luavixen   
2           https://github.com/robe

In [8]:
repo_with_contributor.columns

Index(['owner', 'name_x', 'dependencies_x', 'description_x', 'latestCommit_x',
       'githubLink_x', 'websiteLink_x', 'reputation_x', 'activity_x',
       'popularity', 'languages_x', 'commitCount_x', 'totalStars_x',
       'recentContributors', 'Names', 'name_y', 'dependencies_y', 'email',
       'description_y', 'location', 'organization', 'latestCommit_y',
       'twitterLink', 'githubLink_y', 'websiteLink_y', 'reputation_y',
       'activity_y', 'depActivity', 'languages_y', 'commitCount_y',
       'totalStars_y', 'followers'],
      dtype='object')

In [9]:
r_c = repo_with_contributor.copy()

r_c['activity_y'] = pd.to_numeric(
    r_c['activity_y'].str.replace('%', ''),
    errors='coerce'
)

In [10]:
avg_followers_of_contributor_per_repo = r_c.groupby('githubLink_x')['followers'].mean().reset_index(name='total_avg_followers_by_contributor')

avg_commit_of_contributor_per_repo = r_c.groupby('githubLink_x')['commitCount_y'].mean().reset_index(name='total_avg_commit_by_contributor')

avg_activity_of_contributor_per_repo = r_c.groupby('githubLink_x')['activity_y'].mean().reset_index(name = 'total_avg_activity_by_contributor')

avg_total_stars_of_contributor_per_repo = r_c.groupby('githubLink_x')['totalStars_y'].mean().reset_index(name = 'total_avg_stars_by_contributor')

avg_reputation_of_contributor_per_repo = r_c.groupby('githubLink_x')['reputation_y'].mean().reset_index(name = 'total_avg_reputation_by_contributor')

In [11]:
# Merging all the above dataframes
_temp = pd.merge(
    avg_followers_of_contributor_per_repo,
    avg_commit_of_contributor_per_repo,
    how = 'left',
    on = 'githubLink_x'

).merge(
    avg_activity_of_contributor_per_repo,
    how = 'left',
    on = 'githubLink_x'

).merge(
    avg_total_stars_of_contributor_per_repo,
    how = 'left',
    on = 'githubLink_x'
).merge(
    avg_reputation_of_contributor_per_repo,
    how = 'left',
    on = 'githubLink_x'
)

_temp

githubLink_x  \
0                   https://github.com/101arrowz/fflate   
1                 https://github.com/3-manifolds/cypari   
2                 https://github.com/3-manifolds/fxrays   
3     https://github.com/3-manifolds/knot_floer_homo...   
4                  https://github.com/3-manifolds/plink   
...                                                 ...   
4422                https://github.com/zloirock/core-js   
4423             https://github.com/zombiezen/go-sqlite   
4424   https://github.com/zone117x/ripemd160-minimal.js   
4425                https://github.com/zopfli-rs/zopfli   
4426               https://github.com/zpao/qrcode.react   

      total_avg_followers_by_contributor  total_avg_commit_by_contributor  \
0                                    NaN                              NaN   
1                              13.500000                       130.000000   
2                              17.000000                       131.000000   
3                              17.000000                       131.000000   
4                              13.500000                       130.000000   
...                                  ...                              ...   
4422                         1248.833333                       262.333333   
4423                          150.500000                       196.500000   
4424                                 NaN                              NaN   
4425                          896.000000                       295.500000   
4426                          574.444444                       126.444444   

      total_avg_activity_by_contributor  total_avg_stars_by_contributor  \
0                                   NaN                             NaN   
1                             89.000000                    34574.500000   
2                             89.000000                     1731.000000   
3                             89.000000                     1731.000000   
4                             89.000000                    34574.500000   
...                                 ...                             ...   
4422                          66.500000                    23616.333333   
4423                          77.000000                     4600.250000   
4424                                NaN                             NaN   
4425                          99.000000                    60500.500000   
4426                          69.333333                    83438.000000   

      total_avg_reputation_by_contributor  
0                                     NaN  
1                                6.000000  
2                                5.000000  
3                                5.000000  
4                                6.000000  
...                                   ...  
4422                             6.333333  
4423                             4.250000  
4424                                  NaN  
4425                             7.500000  
4426                             5.222222  

[4427 rows x 6 columns]

In [12]:
enhanced_teams.head()

owner             name dependencies  \
0      oven-sh              bun    no status   
1     zloirock          core-js    no status   
2  paradigmxyz             reth    no status   
3    microsoft       playwright    no status   
4       trezor  trezor-firmware    no status   

                                         description latestCommit  \
0  Incredibly fast JavaScript runtime, bundler, t...     3/1/2025   
1                                   Standard Library     3/1/2025   
2  Modular, contributor-friendly and blazing-fast...     3/1/2025   
3  Playwright is a framework for Web Testing and ...     3/1/2025   
4                    :lock: Trezor Firmware Monorepo    2/28/2025   

                                  githubLink             websiteLink  \
0             https://github.com/oven-sh/bun          https://bun.sh   
1        https://github.com/zloirock/core-js                     NaN   
2        https://github.com/paradigmxyz/reth                     NaN   
3    https://github.com/microsoft/playwright  https://playwright.dev   
4  https://github.com/trezor/trezor-firmware       https://trezor.io   

   reputation activity  popularity  \
0          10      75%          10   
1          12      74%          10   
2           7      74%          10   
3          10      73%          10   
4           6      73%          10   

                                           languages  commitCount  totalStars  \
0         Zig, JavaScript, TypeScript, CSS, Makefile        352.0     76508.0   
1                       JavaScript, HTML, TypeScript        359.0     24860.0   
2  Rust, Dockerfile, Makefile, JavaScript, Handle...        350.0      2722.0   
3           JavaScript, Shell, TypeScript, HTML, CSS        299.0     69737.0   
4               Python, C, Makefile, Nix, Dockerfile        293.0      1437.0   

                                  recentContributors  
0  jprinaldi, luavixen, robertshuford, bjornthepr...  
1  slowcheetah, zloirock, szepeviktor, wchenonly,...  
2  zakisaad, yaziciahmet, thegaram, abnerzheng, m...  
3  aslushnikov, romarionijim, matthijskok, przemy...  
4  gabrielkerekes, mgiannopoulos24, obrusvit, sha...

In [13]:
repo_with_features = pd.merge(
    enhanced_teams,
    _temp,
    left_on = 'githubLink',
    right_on = 'githubLink_x',
    how = 'left'
)

In [14]:
repo_with_features['activity'] = pd.to_numeric(
    repo_with_features['activity'].str.replace('%', ''),
    errors = 'coerce'
)

In [15]:
repo_with_features.head()

owner             name dependencies  \
0      oven-sh              bun    no status   
1     zloirock          core-js    no status   
2  paradigmxyz             reth    no status   
3    microsoft       playwright    no status   
4       trezor  trezor-firmware    no status   

                                         description latestCommit  \
0  Incredibly fast JavaScript runtime, bundler, t...     3/1/2025   
1                                   Standard Library     3/1/2025   
2  Modular, contributor-friendly and blazing-fast...     3/1/2025   
3  Playwright is a framework for Web Testing and ...     3/1/2025   
4                    :lock: Trezor Firmware Monorepo    2/28/2025   

                                  githubLink             websiteLink  \
0             https://github.com/oven-sh/bun          https://bun.sh   
1        https://github.com/zloirock/core-js                     NaN   
2        https://github.com/paradigmxyz/reth                     NaN   
3    https://github.com/microsoft/playwright  https://playwright.dev   
4  https://github.com/trezor/trezor-firmware       https://trezor.io   

   reputation  activity  popularity  \
0          10        75          10   
1          12        74          10   
2           7        74          10   
3          10        73          10   
4           6        73          10   

                                           languages  commitCount  totalStars  \
0         Zig, JavaScript, TypeScript, CSS, Makefile        352.0     76508.0   
1                       JavaScript, HTML, TypeScript        359.0     24860.0   
2  Rust, Dockerfile, Makefile, JavaScript, Handle...        350.0      2722.0   
3           JavaScript, Shell, TypeScript, HTML, CSS        299.0     69737.0   
4               Python, C, Makefile, Nix, Dockerfile        293.0      1437.0   

                                  recentContributors  \
0  jprinaldi, luavixen, robertshuford, bjornthepr...   
1  slowcheetah, zloirock, szepeviktor, wchenonly,...   
2  zakisaad, yaziciahmet, thegaram, abnerzheng, m...   
3  aslushnikov, romarionijim, matthijskok, przemy...   
4  gabrielkerekes, mgiannopoulos24, obrusvit, sha...   

                                githubLink_x  \
0             https://github.com/oven-sh/bun   
1        https://github.com/zloirock/core-js   
2        https://github.com/paradigmxyz/reth   
3    https://github.com/microsoft/playwright   
4  https://github.com/trezor/trezor-firmware   

   total_avg_followers_by_contributor  total_avg_commit_by_contributor  \
0                          167.000000                       243.869955   
1                         1248.833333                       262.333333   
2                           83.311644                       166.565068   
3                          130.116279                       199.434109   
4                           63.194444                       100.416667   

   total_avg_activity_by_contributor  total_avg_stars_by_contributor  \
0                          74.574561                    59013.524664   
1                          66.500000                    23616.333333   
2                          76.847973                     8271.188356   
3                          64.916031                    74487.162791   
4                          70.729730                     7147.750000   

   total_avg_reputation_by_contributor  
0                             5.811404  
1                             6.333333  
2                             4.834459  
3                             6.488550  
4                             4.972973

In [16]:
contributor_string = repo_with_features['recentContributors'].iloc[500]

contributor_list = contributor_string.split(', ')

# len(contributor_list)

In [17]:
big_org.head()

Repository                                  GitHub Link
0        chainsafe/lodestar        https://github.com/chainsafe/lodestar
1       nethereum/nethereum       https://github.com/nethereum/nethereum
2                 wevm/viem                 https://github.com/wevm/viem
3  nethermindeth/nethermind  https://github.com/nethermindeth/nethermind
4         pandas-dev/pandas         https://github.com/pandas-dev/pandas

In [18]:
old_repos['GitHub Link'] = old_repos['poanetwork/hex_fmt'].apply(lambda x: "https://github.com/" + x)

old_repos.head()

poanetwork/hex_fmt                               GitHub Link
0     kevva/astral-regex     https://github.com/kevva/astral-regex
1         mr-tron/base58         https://github.com/mr-tron/base58
2  mitchellh/colorstring  https://github.com/mitchellh/colorstring
3  sile/trackable_derive  https://github.com/sile/trackable_derive
4        indiv0/lazycell        https://github.com/indiv0/lazycell

In [19]:
enhanced_teams.head()

owner             name dependencies  \
0      oven-sh              bun    no status   
1     zloirock          core-js    no status   
2  paradigmxyz             reth    no status   
3    microsoft       playwright    no status   
4       trezor  trezor-firmware    no status   

                                         description latestCommit  \
0  Incredibly fast JavaScript runtime, bundler, t...     3/1/2025   
1                                   Standard Library     3/1/2025   
2  Modular, contributor-friendly and blazing-fast...     3/1/2025   
3  Playwright is a framework for Web Testing and ...     3/1/2025   
4                    :lock: Trezor Firmware Monorepo    2/28/2025   

                                  githubLink             websiteLink  \
0             https://github.com/oven-sh/bun          https://bun.sh   
1        https://github.com/zloirock/core-js                     NaN   
2        https://github.com/paradigmxyz/reth                     NaN   
3    https://github.com/microsoft/playwright  https://playwright.dev   
4  https://github.com/trezor/trezor-firmware       https://trezor.io   

   reputation activity  popularity  \
0          10      75%          10   
1          12      74%          10   
2           7      74%          10   
3          10      73%          10   
4           6      73%          10   

                                           languages  commitCount  totalStars  \
0         Zig, JavaScript, TypeScript, CSS, Makefile        352.0     76508.0   
1                       JavaScript, HTML, TypeScript        359.0     24860.0   
2  Rust, Dockerfile, Makefile, JavaScript, Handle...        350.0      2722.0   
3           JavaScript, Shell, TypeScript, HTML, CSS        299.0     69737.0   
4               Python, C, Makefile, Nix, Dockerfile        293.0      1437.0   

                                  recentContributors  
0  jprinaldi, luavixen, robertshuford, bjornthepr...  
1  slowcheetah, zloirock, szepeviktor, wchenonly,...  
2  zakisaad, yaziciahmet, thegaram, abnerzheng, m...  
3  aslushnikov, romarionijim, matthijskok, przemy...  
4  gabrielkerekes, mgiannopoulos24, obrusvit, sha...

In [20]:
_t = pd.merge(
    test, 
    enhanced_teams,
    left_on = 'repo',
    right_on = 'githubLink',
    how = 'left'
)

_t = pd.merge(
    _t,
    big_org,
    left_on = 'repo',
    right_on = 'GitHub Link',
    how = 'left'
)

_t.shape

no_record = _t[_t['githubLink'].isna() & _t['GitHub Link'].isna()]

no_record_list = no_record['repo'].tolist()


In [21]:
repo_with_features.head()

owner             name dependencies  \
0      oven-sh              bun    no status   
1     zloirock          core-js    no status   
2  paradigmxyz             reth    no status   
3    microsoft       playwright    no status   
4       trezor  trezor-firmware    no status   

                                         description latestCommit  \
0  Incredibly fast JavaScript runtime, bundler, t...     3/1/2025   
1                                   Standard Library     3/1/2025   
2  Modular, contributor-friendly and blazing-fast...     3/1/2025   
3  Playwright is a framework for Web Testing and ...     3/1/2025   
4                    :lock: Trezor Firmware Monorepo    2/28/2025   

                                  githubLink             websiteLink  \
0             https://github.com/oven-sh/bun          https://bun.sh   
1        https://github.com/zloirock/core-js                     NaN   
2        https://github.com/paradigmxyz/reth                     NaN   
3    https://github.com/microsoft/playwright  https://playwright.dev   
4  https://github.com/trezor/trezor-firmware       https://trezor.io   

   reputation  activity  popularity  \
0          10        75          10   
1          12        74          10   
2           7        74          10   
3          10        73          10   
4           6        73          10   

                                           languages  commitCount  totalStars  \
0         Zig, JavaScript, TypeScript, CSS, Makefile        352.0     76508.0   
1                       JavaScript, HTML, TypeScript        359.0     24860.0   
2  Rust, Dockerfile, Makefile, JavaScript, Handle...        350.0      2722.0   
3           JavaScript, Shell, TypeScript, HTML, CSS        299.0     69737.0   
4               Python, C, Makefile, Nix, Dockerfile        293.0      1437.0   

                                  recentContributors  \
0  jprinaldi, luavixen, robertshuford, bjornthepr...   
1  slowcheetah, zloirock, szepeviktor, wchenonly,...   
2  zakisaad, yaziciahmet, thegaram, abnerzheng, m...   
3  aslushnikov, romarionijim, matthijskok, przemy...   
4  gabrielkerekes, mgiannopoulos24, obrusvit, sha...   

                                githubLink_x  \
0             https://github.com/oven-sh/bun   
1        https://github.com/zloirock/core-js   
2        https://github.com/paradigmxyz/reth   
3    https://github.com/microsoft/playwright   
4  https://github.com/trezor/trezor-firmware   

   total_avg_followers_by_contributor  total_avg_commit_by_contributor  \
0                          167.000000                       243.869955   
1                         1248.833333                       262.333333   
2                           83.311644                       166.565068   
3                          130.116279                       199.434109   
4                           63.194444                       100.416667   

   total_avg_activity_by_contributor  total_avg_stars_by_contributor  \
0                          74.574561                    59013.524664   
1                          66.500000                    23616.333333   
2                          76.847973                     8271.188356   
3                          64.916031                    74487.162791   
4                          70.729730                     7147.750000   

   total_avg_reputation_by_contributor  
0                             5.811404  
1                             6.333333  
2                             4.834459  
3                             6.488550  
4                             4.972973

In [22]:
all_repos_with_features = repo_with_features.copy()

In [23]:
_test = pd.merge(
    test,
    all_repos_with_features,
    left_on = 'repo',
    right_on = 'githubLink',
    how = 'left'
)

_test

repo    parent  \
0                      https://github.com/a16z/helios  ethereum   
1                   https://github.com/alloy-rs/alloy  ethereum   
2                      https://github.com/apeworx/ape  ethereum   
3               https://github.com/chainsafe/lodestar  ethereum   
4                   https://github.com/consensys/teku  ethereum   
5                https://github.com/erigontech/erigon  ethereum   
6   https://github.com/eth-infinitism/account-abst...  ethereum   
7            https://github.com/ethereum-lists/chains  ethereum   
8         https://github.com/ethereum/consensus-specs  ethereum   
9                    https://github.com/ethereum/eips  ethereum   
10                 https://github.com/ethereum/evmone  ethereum   
11         https://github.com/ethereum/execution-apis  ethereum   
12                     https://github.com/argotorg/fe  ethereum   
13            https://github.com/ethereum/go-ethereum  ethereum   
14                 https://github.com/ethereum/py-evm  ethereum   
15          https://github.com/ethereum/remix-project  ethereum   
16               https://github.com/argotorg/solidity  ethereum   
17               https://github.com/argotorg/sourcify  ethereum   
18                https://github.com/ethereum/web3.py  ethereum   
19             https://github.com/ethers-io/ethers.js  ethereum   
20              https://github.com/foundry-rs/foundry  ethereum   
21           https://github.com/grandinetech/grandine  ethereum   
22         https://github.com/hyperledger-web3j/web3j  ethereum   
23                https://github.com/hyperledger/besu  ethereum   
24             https://github.com/nethereum/nethereum  ethereum   
25        https://github.com/nethermindeth/nethermind  ethereum   
26         https://github.com/nomicfoundation/hardhat  ethereum   
27  https://github.com/openzeppelin/openzeppelin-c...  ethereum   
28                https://github.com/paradigmxyz/reth  ethereum   
29             https://github.com/prysmaticlabs/prysm  ethereum   
30  https://github.com/safe-global/safe-smart-account  ethereum   
31     https://github.com/scaffold-eth/scaffold-eth-2  ethereum   
32                 https://github.com/sigp/lighthouse  ethereum   
33           https://github.com/status-im/nimbus-eth2  ethereum   
34             https://github.com/vyperlang/titanoboa  ethereum   
35                 https://github.com/vyperlang/vyper  ethereum   
36                       https://github.com/wevm/viem  ethereum   
37                   https://github.com/argotorg/hevm  ethereum   
38                 https://github.com/ethdebug/format  ethereum   
39                    https://github.com/argotorg/act  ethereum   
40    https://github.com/ethpandaops/ethereum-package  ethereum   
41  https://github.com/ethpandaops/ethereum-helm-c...  ethereum   
42         https://github.com/ethpandaops/checkpointz  ethereum   
43  https://github.com/lambdaclass/lambda_ethereum...  ethereum   
44             https://github.com/erigontech/silkworm  ethereum   

                owner                    name dependencies  \
0                a16z                  helios    no status   
1            alloy-rs                   alloy    no status   
2                 NaN                     NaN          NaN   
3                 NaN                     NaN          NaN   
4                 NaN                     NaN          NaN   
5                 NaN                     NaN          NaN   
6      eth-infinitism     account-abstraction    no status   
7                 NaN                     NaN          NaN   
8                 NaN                     NaN          NaN   
9                 NaN                     NaN          NaN   
10                NaN                     NaN          NaN   
11                NaN                     NaN          NaN   
12                NaN                     NaN          NaN   
13           ethereum             go-ethereum    no status   
14           ethereum               

In [24]:
# repo_with_features[repo_with_features['githubLink'].notna() & repo_with_features['githubLink_x'].isna()]

# repo_with_features['githubLink_x'].isna().sum()

In [25]:
no_record_list[0]

'https://github.com/ethereum/consensus-specs'

In [26]:
_t.head(5)

repo    parent     owner    name  \
0         https://github.com/a16z/helios  ethereum      a16z  helios   
1      https://github.com/alloy-rs/alloy  ethereum  alloy-rs   alloy   
2         https://github.com/apeworx/ape  ethereum       NaN     NaN   
3  https://github.com/chainsafe/lodestar  ethereum       NaN     NaN   
4      https://github.com/consensys/teku  ethereum       NaN     NaN   

  dependencies                                        description  \
0    no status  A fast, secure, and portable multichain light ...   
1    no status  Transports, Middleware, and Networks for the A...   
2          NaN                                                NaN   
3          NaN                                                NaN   
4          NaN                                                NaN   

  latestCommit                         githubLink       websiteLink  \
0    2/27/2025     https://github.com/a16z/helios               NaN   
1     3/1/2025  https://github.com/alloy-rs/alloy  https://alloy.rs   
2          NaN                                NaN               NaN   
3          NaN                                NaN               NaN   
4          NaN                                NaN               NaN   

   reputation activity  popularity                                  languages  \
0         8.0      61%        10.0  Rust, Shell, HTML, TypeScript, JavaScript   
1         7.0      72%        10.0                                Rust, Shell   
2         NaN      NaN         NaN                                        NaN   
3         NaN      NaN         NaN                                        NaN   
4         NaN      NaN         NaN                                        NaN   

   commitCount  totalStars                                 recentContributors  \
0         63.0      1948.0  eshaan7, ncitron, xavierdmello, karen-sarkisya...   
1        310.0       750.0  neithanmo, abnerzheng, fgimenez, brechtpd, ste...   
2          NaN         NaN                                                NaN   
3          NaN         NaN                                                NaN   
4          NaN         NaN                                                NaN   

           Repository                            GitHub Link  
0                 NaN                                    NaN  
1                 NaN                                    NaN  
2         apeworx/ape         https://github.com/apeworx/ape  
3  chainsafe/lodestar  https://github.com/chainsafe/lodestar  
4      consensys/teku      https://github.com/consensys/teku

In [ ]:
import base64
import requests
import time

GITHUB_TOKEN = "" 

HEADERS = {'Authorization': f'token {GITHUB_TOKEN}'}

test_url = 'https://api.github.com/repos/facebook/react'

response = requests.get(test_url, headers=HEADERS)

# print(f"Status code for React repo: {response.status_code}")


In [28]:
test_url = "https://github.com/ethpandaops/ethereum-package"	
response = requests.get(test_url, headers = HEADERS)

print(response.status_code)

200


In [29]:
no_record_list

['https://github.com/ethereum/consensus-specs',
 'https://github.com/ethereum/eips',
 'https://github.com/ethereum/evmone',
 'https://github.com/ethereum/execution-apis',
 'https://github.com/argotorg/fe',
 'https://github.com/argotorg/solidity',
 'https://github.com/argotorg/sourcify',
 'https://github.com/argotorg/hevm',
 'https://github.com/ethdebug/format',
 'https://github.com/argotorg/act',
 'https://github.com/ethpandaops/ethereum-package',
 'https://github.com/ethpandaops/ethereum-helm-charts',
 'https://github.com/ethpandaops/checkpointz',
 'https://github.com/lambdaclass/lambda_ethereum_consensus',
 'https://github.com/erigontech/silkworm']

In [30]:
import os
from getpass import getpass

headers = {
    'Accept': 'application/vnd.github.v3+json',
    'Authorization': f'token {GITHUB_TOKEN}'
}

# Test the authentication
test_response = requests.get('https://api.github.com/rate_limit', headers=headers)
if test_response.status_code == 200:
    limits = test_response.json()['rate']
    print(f"Authentication successful!")
    print(f"Rate limit: {limits['limit']}")
    print(f"Remaining: {limits['remaining']}")
    print(f"Reset time: {pd.Timestamp.fromtimestamp(limits['reset'])}")
else:
    print("Authentication failed!")
    print(test_response.json())



Authentication successful!
Rate limit: 5000
Remaining: 4831
Reset time: 2025-09-14 23:03:38


In [ ]:
MISTRAL_API = "<REPLACE_WITH_YOUR_MISTRAL_API_KEY>"

GEMINI_API_KEY = "<REPLACE_WITH_YOUR_GEMINI_API_KEY>"

In [32]:
# stars = repo_data.get('stargazers_count', 'N/A')
# forks = repo_data.get('forks_count', 'N/A')
# open_issues = repo_data.get('open_issues_count', 'N/A')
# last_push = repo_data.get('pushed_at', 'N/A')

In [33]:
# Loop through each repo URL in no_record_list and print the number of stars for each
for _url in no_record_list:
    # Extract owner and repo from the URL
    try:
        _, _, _, owner, repo = _url.rstrip('/').split('/')
    except ValueError:
        print(f"Could not parse owner/repo from URL: {_url}")
        continue

    api_url = f'https://api.github.com/repos/{owner}/{repo}'
    response = requests.get(api_url, headers=headers)

    if response.status_code == 200:
        repo_data = response.json()
        stars = repo_data.get('stargazers_count', 'N/A')
        print(f"{owner}/{repo}: {stars} stars")
    else:
        print(f"Failed to fetch data for {owner}/{repo}: {response.status_code}")

ethereum/consensus-specs: 3783 stars
ethereum/eips: 13507 stars
ethereum/evmone: 948 stars
ethereum/execution-apis: 1060 stars
argotorg/fe: 1683 stars
argotorg/solidity: 25138 stars
argotorg/sourcify: 866 stars
argotorg/hevm: 305 stars
ethdebug/format: 69 stars
argotorg/act: 247 stars
ethpandaops/ethereum-package: 397 stars
ethpandaops/ethereum-helm-charts: 178 stars
ethpandaops/checkpointz: 155 stars
lambdaclass/lambda_ethereum_consensus: 108 stars
erigontech/silkworm: 306 stars


In [34]:
all_repos_with_features

owner             name dependencies  \
0                    oven-sh              bun    no status   
1                   zloirock          core-js    no status   
2                paradigmxyz             reth    no status   
3                  microsoft       playwright    no status   
4                     trezor  trezor-firmware    no status   
...                      ...              ...          ...   
4422                  coreos       go-systemd    no status   
4423             cockroachdb           errors    no status   
4424              manifoldco         promptui    no status   
4425             cockroachdb          logtags    no status   
4426  census-instrumentation    opencensus-go    no status   

                                            description latestCommit  \
0     Incredibly fast JavaScript runtime, bundler, t...     3/1/2025   
1                                      Standard Library     3/1/2025   
2     Modular, contributor-friendly and blazing-fast...     3/1/2025   
3     Playwright is a framework for Web Testing and ...     3/1/2025   
4                       :lock: Trezor Firmware Monorepo    2/28/2025   
...                                                 ...          ...   
4422  Go bindings to systemd socket activation, jour...   no commits   
4423  Go error library with error portability over t...     6/3/2024   
4424   Interactive prompt for command-line applications   no commits   
4425              key/value annotations for Go contexts   12/15/2024   
4426  A stats collection and distributed tracing fra...   no commits   

                                             githubLink  \
0                        https://github.com/oven-sh/bun   
1                   https://github.com/zloirock/core-js   
2                   https://github.com/paradigmxyz/reth   
3               https://github.com/microsoft/playwright   
4             https://github.com/trezor/trezor-firmware   
...                                                 ...   
4422               https://github.com/coreos/go-systemd   
4423              https://github.com/cockroachdb/errors   
4424             https://github.com/manifoldco/promptui   
4425             https://github.com/cockroachdb/logtags   
4426  https://github.com/census-instrumentation/open...   

                  websiteLink  reputation  activity  popularity  \
0              https://bun.sh          10        75          10   
1                         NaN          12        74          10   
2                         NaN           7        74          10   
3      https://playwright.dev          10        73          10   
4           https://trezor.io           6        73          10   
...                       ...         ...       ...         ...   
4422                      NaN           6         0          10   
4423                      NaN           8         0          10   
4424  https://www.manifold.co           5         0          10   
4425                      NaN           8         0           4   
4426     http://opencensus.io           6         0          10   

                                              languages  commitCount  \
0            Zig, JavaScript, TypeScript, CSS, Makefile        352.0   
1                          JavaScript, HTML, TypeScript        359.0   
2     Rust, Dockerfile, Makefile, JavaScript, Handle...        350.0   
3              JavaScript, Shell, TypeScript, HTML, CSS        299.0   
4                  Python, C, Makefile, Nix, Dockerfile        293.0   
...                                                 ...          ...   
4422                                          Go, Shell          0.0   
4423                                                 Go          3.0   
4424                                       Go, Makefile          0.0   
4425                                                 Go          2.0   
4426                          Go, Shell, HTML, Makefile          0.0   

      totalStars                  

In [35]:
_train = pd.merge(
    train,
    all_repos_with_features,
    left_on = 'repo_a',
    right_on = 'githubLink',
    how = 'left'
)

_train = pd.merge(
    _train,
    all_repos_with_features,
    left_on = 'repo_b',
    right_on = 'githubLink',
    how = 'left'
)

_train

timestamp      juror  \
0    2025-01-24T19:03:52.628Z   L1Juror1   
1    2025-01-24T19:20:19.601Z   L1Juror1   
2    2025-01-15T11:19:51.121Z  L1Juror10   
3    2025-01-15T11:22:16.292Z  L1Juror10   
4    2025-01-15T11:24:06.402Z  L1Juror10   
..                        ...        ...   
121  2025-03-01T07:19:51.280Z   L1Juror9   
122  2025-03-01T07:23:49.752Z   L1Juror9   
123  2025-03-01T07:31:41.353Z   L1Juror9   
124  2025-03-01T07:34:19.636Z   L1Juror9   
125  2025-03-01T07:35:55.487Z   L1Juror9   

                                                repo_a  \
0                 https://github.com/erigontech/erigon   
1                 https://github.com/erigontech/erigon   
2                 https://github.com/erigontech/erigon   
3               https://github.com/prysmaticlabs/prysm   
4    https://github.com/safe-global/safe-smart-account   
..                                                 ...   
121                  https://github.com/consensys/teku   
122  https://github.com/safe-global/safe-smart-account   
123              https://github.com/foundry-rs/foundry   
124          https://github.com/ethereum/remix-project   
125           https://github.com/ethereum-lists/chains   

                                      repo_b    parent  choice  multiplier  \
0        https://github.com/hyperledger/besu  ethereum     2.0        5.27   
1    https://github.com/ethereum/go-ethereum  ethereum     2.0       10.55   
2        https://github.com/ethereum/web3.py  ethereum     1.0       50.00   
3        https://github.com/hyperledger/besu  ethereum     1.0        3.00   
4        https://github.com/paradigmxyz/reth  ethereum     1.0        1.50   
..                                       ...       ...     ...         ...   
121      https://github.com/ethereum/web3.py  ethereum     2.0       10.00   
122      https://github.com/paradigmxyz/reth  ethereum     2.0        8.00   
123      https://github.com/paradigmxyz/reth  ethereum     1.0       20.00   
124   https://github.com/prysmaticlabs/prysm  ethereum     2.0       10.00   
125      https://github.com/ethereum/web3.py  ethereum     2.0       10.00   

     reasoning        owner_x              name_x  ...  \
0          NaN            NaN                 NaN  ...   
1          NaN            NaN                 NaN  ...   
2          NaN            NaN                 NaN  ...   
3          NaN  prysmaticlabs               prysm  ...   
4          NaN    safe-global  safe-smart-account  ...   
..         ...            ...                 ...  ...   
121        NaN            NaN                 NaN  ...   
122        NaN    safe-global  safe-smart-account  ...   
123        NaN     foundry-rs             foundry  ...   
124        NaN       ethereum       remix-project  ...   
125        NaN            NaN                 NaN  ...   

                                           languages_y commitCount_y  \
0        JavaScript, Java, Shell, Dockerfile, Solidity         253.0   
1                         Go, Java, Shell, C, Makefile         241.0   
2        Makefile, Python, Dockerfile, Solidity, Shell         165.0   
3        JavaScript, Java, Shell, Dockerfile, Solidity         253.0   
4    Rust, Dockerfile, Makefile, JavaScript, Handle...         350.0   
..                                                 ...           ...   
121      Makefile, Python, Dockerfile, Solidity, Shell         165.0   
122  Rust, Dockerfile, Makefile, JavaScript, Handle...         350.0   
123  Rust, Dockerfile, Makefile, JavaScript, Handle...         350.0   
124            Go, Shell, Dockerfile, Starlark, Smarty         247.0   
125      Makefile, Python, Dockerfile, Solidity, Shell         165.0   

    totalStars_y                               recentContributors_y  \
0         1483.0  amsmota, tiesd, gabriel-trintinalia, skylenet,...   
1        47255.0  cedrick-ah, decanus, qcrao, gzliudan, marcinds...   
2         4967.0  defidebauchery, davidromanovizc, lrusak, darwi...   
3         1

In [36]:
_train.columns

Index(['timestamp', 'juror', 'repo_a', 'repo_b', 'parent', 'choice',
       'multiplier', 'reasoning', 'owner_x', 'name_x', 'dependencies_x',
       'description_x', 'latestCommit_x', 'githubLink_x', 'websiteLink_x',
       'reputation_x', 'activity_x', 'popularity_x', 'languages_x',
       'commitCount_x', 'totalStars_x', 'recentContributors_x',
       'githubLink_x_x', 'total_avg_followers_by_contributor_x',
       'total_avg_commit_by_contributor_x',
       'total_avg_activity_by_contributor_x',
       'total_avg_stars_by_contributor_x',
       'total_avg_reputation_by_contributor_x', 'owner_y', 'name_y',
       'dependencies_y', 'description_y', 'latestCommit_y', 'githubLink_y',
       'websiteLink_y', 'reputation_y', 'activity_y', 'popularity_y',
       'languages_y', 'commitCount_y', 'totalStars_y', 'recentContributors_y',
       'githubLink_x_y', 'total_avg_followers_by_contributor_y',
       'total_avg_commit_by_contributor_y',
       'total_avg_activity_by_contributor_y',
    

In [37]:
completely_missing_train = _train[_train['githubLink_x'].isna() & _train['githubLink_y'].isna()]

completely_missing_train

timestamp      juror  \
16   2025-01-23T16:59:25.720Z  L1Juror13   
18   2025-01-20T10:07:56.392Z  L1Juror14   
27   2025-03-01T16:08:01.124Z  L1Juror15   
60   2025-03-30T05:45:07.269Z  L1Juror24   
62   2025-03-30T07:19:50.366Z  L1Juror24   
64   2025-03-30T07:25:51.954Z  L1Juror24   
80   2025-03-30T04:24:26.983Z  L1Juror28   
83   2025-03-30T05:01:54.338Z  L1Juror28   
89   2025-03-30T21:18:25.091Z  L1Juror29   
98   2025-03-30T21:27:35.282Z  L1Juror29   
115  2025-02-10T10:22:16.507Z   L1Juror6   
117  2025-01-06T14:31:20.921Z   L1Juror7   

                                          repo_a  \
16          https://github.com/erigontech/erigon   
18          https://github.com/argotorg/solidity   
27             https://github.com/consensys/teku   
60          https://github.com/argotorg/sourcify   
62   https://github.com/nethermindeth/nethermind   
64             https://github.com/consensys/teku   
80      https://github.com/ethereum-lists/chains   
83          https://github.com/argotorg/sourcify   
89             https://github.com/consensys/teku   
98          https://github.com/argotorg/sourcify   
115         https://github.com/erigontech/erigon   
117        https://github.com/chainsafe/lodestar   

                                          repo_b    parent  choice  \
16      https://github.com/ethereum-lists/chains  ethereum     1.0   
18   https://github.com/nethermindeth/nethermind  ethereum     1.0   
27          https://github.com/erigontech/erigon  ethereum     1.0   
60          https://github.com/erigontech/erigon  ethereum     2.0   
62        https://github.com/nethereum/nethereum  ethereum     1.0   
64          https://github.com/argotorg/sourcify  ethereum     1.0   
80        https://github.com/nethereum/nethereum  ethereum     2.0   
83             https://github.com/consensys/teku  ethereum     2.0   
89          https://github.com/erigontech/erigon  ethereum     2.0   
98        https://github.com/nethereum/nethereum  ethereum     2.0   
115                 https://github.com/wevm/viem  ethereum     1.0   
117            https://github.com/consensys/teku  ethereum     2.0   

     multiplier  reasoning owner_x name_x  ... languages_y commitCount_y  \
16          5.0        NaN     NaN    NaN  ...         NaN           NaN   
18         10.0        NaN     NaN    NaN  ...         NaN           NaN   
27          2.5        NaN     NaN    NaN  ...         NaN           NaN   
60         10.0        NaN     NaN    NaN  ...         NaN           NaN   
62         10.0        NaN     NaN    NaN  ...         NaN           NaN   
64         20.0        NaN     NaN    NaN  ...         NaN           NaN   
80          2.0        NaN     NaN    NaN  ...         NaN           NaN   
83          5.0        NaN     NaN    NaN  ...         NaN           NaN   
89          1.0        NaN     NaN    NaN  ...         NaN           NaN   
98          2.0        NaN     NaN    NaN  ...         NaN           NaN   
115         2.0        NaN     NaN    NaN  ...         NaN           NaN   
117         1.5        NaN     NaN    NaN  ...         NaN           NaN   

    totalStars_y recentContributors_y githubLink_x_y  \
16           NaN                  NaN            NaN   
18           NaN                  NaN            NaN   
27           NaN                  NaN            NaN   
60           NaN                  NaN            NaN   
62           NaN                  NaN            NaN   
64           NaN                  NaN            NaN   
80           NaN                  NaN            NaN   
83           NaN                  NaN            NaN   
89           NaN                  NaN            NaN   
98           NaN                  NaN            NaN   
115          NaN                  NaN            NaN   
117          NaN                  NaN            NaN   

     total_avg_followers_by_contributor_y  total_avg_commit_by_contributor_y  \
16                                    NaN                    

In [38]:
complete_training_information = _train[_train['githubLink_x'].notna() & _train['githubLink_y'].notna()]

partial_training_information = _train[
    (_train['githubLink_x'].isna() & _train['githubLink_y'].notna()) |
    (_train['githubLink_x'].notna() & _train['githubLink_y'].isna())
]

partial_training_information.shape

(56, 48)

In [39]:
missing_repos = set()

missing_repos.update(
    partial_training_information.loc[
        partial_training_information['githubLink_x'].isna() & partial_training_information['githubLink_y'].notna(),
        'repo_a'
    ]
)

missing_repos.update(
    partial_training_information.loc[
        partial_training_information['githubLink_y'].isna() & partial_training_information['githubLink_x'].notna(),
        'repo_b'
    ]
)

missing_repos_in_big_org = set(big_org['GitHub Link']).intersection(missing_repos)

print(f"Total repos with missing features on one side: {len(missing_repos)}")
print(f"Of these, number present in big_org: {len(missing_repos_in_big_org)}")
print("Missing repos in big_org:", missing_repos_in_big_org)

Total repos with missing features on one side: 10
Of these, number present in big_org: 7
Missing repos in big_org: {'https://github.com/nethermindeth/nethermind', 'https://github.com/chainsafe/lodestar', 'https://github.com/erigontech/erigon', 'https://github.com/wevm/viem', 'https://github.com/nethereum/nethereum', 'https://github.com/ethereum-lists/chains', 'https://github.com/consensys/teku'}


Out of 126 rows in original train, after merging with all_repos_with_feature the new _train contains 58 rows with almost all the information (haven't check anything besides 'githubLink_x', 'githubLink_y') , 29 and 27 are the numbers where either one of them present but the second is not, while 12 are those where none of the information is present.

In [40]:
_ = set(missing_repos - missing_repos_in_big_org).intersection(old_repos['GitHub Link'])

# _

missing_repos - missing_repos_in_big_org

old_repos[old_repos['GitHub Link'] == 'https://github.com/argotorg/sourcify']

Empty DataFrame
Columns: [poanetwork/hex_fmt, GitHub Link]
Index: []

In [41]:
_big_train = completely_missing_train[['repo_a', 'repo_b']]

_big_train = pd.merge(
    _big_train,
    big_org,
    left_on = 'repo_a',
    right_on = 'GitHub Link',
    how = 'left'
)

_big_train = pd.merge(
    _big_train,
    big_org,
    left_on = 'repo_b',
    right_on = 'GitHub Link',
    how = 'left'
)

_big_train



repo_a  \
0          https://github.com/erigontech/erigon   
1          https://github.com/argotorg/solidity   
2             https://github.com/consensys/teku   
3          https://github.com/argotorg/sourcify   
4   https://github.com/nethermindeth/nethermind   
5             https://github.com/consensys/teku   
6      https://github.com/ethereum-lists/chains   
7          https://github.com/argotorg/sourcify   
8             https://github.com/consensys/teku   
9          https://github.com/argotorg/sourcify   
10         https://github.com/erigontech/erigon   
11        https://github.com/chainsafe/lodestar   

                                         repo_b              Repository_x  \
0      https://github.com/ethereum-lists/chains         erigontech/erigon   
1   https://github.com/nethermindeth/nethermind                       NaN   
2          https://github.com/erigontech/erigon            consensys/teku   
3          https://github.com/erigontech/erigon                       NaN   
4        https://github.com/nethereum/nethereum  nethermindeth/nethermind   
5          https://github.com/argotorg/sourcify            consensys/teku   
6        https://github.com/nethereum/nethereum     ethereum-lists/chains   
7             https://github.com/consensys/teku                       NaN   
8          https://github.com/erigontech/erigon            consensys/teku   
9        https://github.com/nethereum/nethereum                       NaN   
10                 https://github.com/wevm/viem         erigontech/erigon   
11            https://github.com/consensys/teku        chainsafe/lodestar   

                                  GitHub Link_x              Repository_y  \
0          https://github.com/erigontech/erigon     ethereum-lists/chains   
1                                           NaN  nethermindeth/nethermind   
2             https://github.com/consensys/teku         erigontech/erigon   
3                                           NaN         erigontech/erigon   
4   https://github.com/nethermindeth/nethermind       nethereum/nethereum   
5             https://github.com/consensys/teku                       NaN   
6      https://github.com/ethereum-lists/chains       nethereum/nethereum   
7                                           NaN            consensys/teku   
8             https://github.com/consensys/teku         erigontech/erigon   
9                                           NaN       nethereum/nethereum   
10         https://github.com/erigontech/erigon                 wevm/viem   
11        https://github.com/chainsafe/lodestar            consensys/teku   

                                  GitHub Link_y  
0      https://github.com/ethereum-lists/chains  
1   https://github.com/nethermindeth/nethermind  
2          https://github.com/erigontech/erigon  
3          https://github.com/erigontech/erigon  
4        https://github.com/nethereum/nethereum  
5                                           NaN  
6        https://github.com/nethereum/nethereum  
7             https://github.com/consensys/teku  
8          https://github.com/erigontech/erigon  
9        https://github.com/nethereum/nethereum  
10                 https://github.com/wevm/viem  
11            https://github.com/consensys/teku

In [42]:
# Collect all matched repos from repo_a and repo_b where there is a match in big_org
matched_repos = set()

matched_repos.update(_big_train.loc[_big_train['GitHub Link_x'].notna(), 'repo_a'])

matched_repos.update(_big_train.loc[_big_train['GitHub Link_y'].notna(), 'repo_b'])

big_org_in_train = list(matched_repos)

big_org_in_train

['https://github.com/nethermindeth/nethermind',
 'https://github.com/chainsafe/lodestar',
 'https://github.com/erigontech/erigon',
 'https://github.com/wevm/viem',
 'https://github.com/nethereum/nethereum',
 'https://github.com/ethereum-lists/chains',
 'https://github.com/consensys/teku']

In [43]:
missing_repos_in_big_org

{'https://github.com/chainsafe/lodestar',
 'https://github.com/consensys/teku',
 'https://github.com/erigontech/erigon',
 'https://github.com/ethereum-lists/chains',
 'https://github.com/nethereum/nethereum',
 'https://github.com/nethermindeth/nethermind',
 'https://github.com/wevm/viem'}

Analyizng the completely missing thing above and mapping it to the big_org, we can trace back each of the repos, but big_repos doesn't contain much of the information either.

In [44]:
# Trying to fetch the stars and other features till the data of publishing data, i.e., 31st jan 2025.


In [45]:
url = big_org_in_train[0]
readme_content = []
_, _, _, owner, repo = url.rstrip('/').split('/')
names = ['README.md', 'Readme.md', 'readme.md', 'README.markdown']

for name in names:
    _url = f'https://api.github.com/repos/{owner}/{repo}/contents/{name}'
    response = requests.get(_url, headers = headers)

    if response.status_code == 200:
        content = base64.b64decode(response.json()['content']).decode('utf-8')
        readme_content.append(content)

In [46]:
readme_content

['<p align="center">\n  <picture>\n    <source media="(prefers-color-scheme: dark)" srcset="https://github.com/nethermindeth/nethermind/assets/337518/3e3b3c06-9cf3-4364-a774-158e649588cc">\n    <source media="(prefers-color-scheme: light)" srcset="https://github.com/nethermindeth/nethermind/assets/337518/d1cc365c-6045-409f-a961-18d22ddb2535">\n    <img alt="Nethermind" src="https://github.com/nethermindeth/nethermind/assets/337518/d1cc365c-6045-409f-a961-18d22ddb2535" height="64">\n  </picture>\n</p>\n\n# Nethermind Ethereum client\n\n[![Tests](https://github.com/nethermindeth/nethermind/actions/workflows/nethermind-tests.yml/badge.svg)](https://github.com/nethermindeth/nethermind/actions/workflows/nethermind-tests.yml)\n[![Follow us on X](https://img.shields.io/twitter/follow/nethermindeth?style=social&label=Follow%20us)](https://x.com/nethermindeth)\n[![Chat on Discord](https://img.shields.io/discord/629004402170134531?style=social&logo=discord)](https://discord.gg/GXJFaYk)\n[![GitHu

In [47]:
# from typing import List
# import requests
# import base64
# import pandas as pd

# def fetching_readme_and_description(repo_urls: List[str]) -> pd.DataFrame:
#     rows = []
#     for url in repo_urls:
#         try:
#             _, _, _, owner, repo = url.rstrip('/').split('/')
#         except ValueError:
#             # Skip malformed URLs
#             continue

#         # Fetch README content
#         names = ['README.md', 'Readme.md', 'readme.md', 'README.markdown']
#         readme_content = None
#         for name in names:
#             _url = f'https://api.github.com/repos/{owner}/{repo}/contents/{name}'
#             response = requests.get(_url, headers=headers)
#             if response.status_code == 200:
#                 try:
#                     content = base64.b64decode(response.json()['content']).decode('utf-8')
#                 except Exception:
#                     content = None
#                 readme_content = content
#                 break  # Stop after finding the first available README

#         # Fetch repo description/about
#         repo_api_url = f'https://api.github.com/repos/{owner}/{repo}'
#         desc_response = requests.get(repo_api_url, headers=headers)
#         if desc_response.status_code == 200:
#             description = desc_response.json().get('description', None)
#             stars = desc_response.json().get('stargazers_count', None)
#             commits 
            
#         else:
#             description = None
        
         

#         rows.append({'repo': url,'totalStars': stars, 'readme_content': readme_content, 'description': description})
#     return pd.DataFrame(rows)

# big_org_features = fetching_readme_and_description(big_org_in_train)
# big_org_features


In [48]:
# big_org_features.iloc[0]['readme_content']

In [49]:
# finding the common repos in big_org 

In [50]:
# There are no common columns between big_org and enhanced_teams or repo_with_features

big_org_features = pd.merge(
    big_org,
    all_repos_with_features,
    left_on = 'GitHub Link',
    right_on = 'githubLink',
    how = 'left'
)

big_org_features.head(3)

Repository                             GitHub Link owner name  \
0   chainsafe/lodestar   https://github.com/chainsafe/lodestar   NaN  NaN   
1  nethereum/nethereum  https://github.com/nethereum/nethereum   NaN  NaN   
2            wevm/viem            https://github.com/wevm/viem   NaN  NaN   

  dependencies description latestCommit githubLink websiteLink  reputation  \
0          NaN         NaN          NaN        NaN         NaN         NaN   
1          NaN         NaN          NaN        NaN         NaN         NaN   
2          NaN         NaN          NaN        NaN         NaN         NaN   

   ...  languages  commitCount totalStars  recentContributors  githubLink_x  \
0  ...        NaN          NaN        NaN                 NaN           NaN   
1  ...        NaN          NaN        NaN                 NaN           NaN   
2  ...        NaN          NaN        NaN                 NaN           NaN   

  total_avg_followers_by_contributor total_avg_commit_by_contributor  \
0                                NaN                             NaN   
1                                NaN                             NaN   
2                                NaN                             NaN   

   total_avg_activity_by_contributor  total_avg_stars_by_contributor  \
0                                NaN                             NaN   
1                                NaN                             NaN   
2                                NaN                             NaN   

   total_avg_reputation_by_contributor  
0                                  NaN  
1                                  NaN  
2                                  NaN  

[3 rows x 22 columns]

In [51]:
# Again there are no common columns between old_repos and enhanced_teams or repo_with_features

old_repos_features = pd.merge(
    old_repos,
    all_repos_with_features,
    left_on = 'GitHub Link',
    right_on = 'githubLink',
    how = 'left'
)

old_repos_features.head()

# old_repos_features.isna().sum()

poanetwork/hex_fmt                               GitHub Link owner name  \
0     kevva/astral-regex     https://github.com/kevva/astral-regex   NaN  NaN   
1         mr-tron/base58         https://github.com/mr-tron/base58   NaN  NaN   
2  mitchellh/colorstring  https://github.com/mitchellh/colorstring   NaN  NaN   
3  sile/trackable_derive  https://github.com/sile/trackable_derive   NaN  NaN   
4        indiv0/lazycell        https://github.com/indiv0/lazycell   NaN  NaN   

  dependencies description latestCommit githubLink websiteLink  reputation  \
0          NaN         NaN          NaN        NaN         NaN         NaN   
1          NaN         NaN          NaN        NaN         NaN         NaN   
2          NaN         NaN          NaN        NaN         NaN         NaN   
3          NaN         NaN          NaN        NaN         NaN         NaN   
4          NaN         NaN          NaN        NaN         NaN         NaN   

   ...  languages  commitCount totalStars  recentContributors  githubLink_x  \
0  ...        NaN          NaN        NaN                 NaN           NaN   
1  ...        NaN          NaN        NaN                 NaN           NaN   
2  ...        NaN          NaN        NaN                 NaN           NaN   
3  ...        NaN          NaN        NaN                 NaN           NaN   
4  ...        NaN          NaN        NaN                 NaN           NaN   

  total_avg_followers_by_contributor total_avg_commit_by_contributor  \
0                                NaN                             NaN   
1                                NaN                             NaN   
2                                NaN                             NaN   
3                                NaN                             NaN   
4                                NaN                             NaN   

   total_avg_activity_by_contributor  total_avg_stars_by_contributor  \
0                                NaN                             NaN   
1                                NaN                             NaN   
2                                NaN                             NaN   
3                                NaN                             NaN   
4                                NaN                             NaN   

   total_avg_reputation_by_contributor  
0                                  NaN  
1                                  NaN  
2                                  NaN  
3                                  NaN  
4                                  NaN  

[5 rows x 22 columns]

In [52]:
enhanced_teams.columns

Index(['owner', 'name', 'dependencies', 'description', 'latestCommit',
       'githubLink', 'websiteLink', 'reputation', 'activity', 'popularity',
       'languages', 'commitCount', 'totalStars', 'recentContributors'],
      dtype='object')

In [53]:
readme_content = []
for url in no_record_list:
    _, _, _, owner, repo = url.rstrip('/').split('/')
    names = ['README.md', 'Readme.md', 'readme.md', 'README.markdown']

    for name in names:
        _url = f'https://api.github.com/repos/{owner}/{repo}/contents/{name}'
        response = requests.get(_url, headers = headers)

        if response.status_code == 200:
            content = base64.b64decode(response.json()['content']).decode('utf-8')
            readme_content.append(content)
            # print(content)

# print(fetch_count)

In [54]:
readme_content[0]

'# Ethereum Proof-of-Stake Consensus Specifications\n\n[![Join the chat at https://discord.gg/qGpsxSA](https://img.shields.io/badge/chat-on%20discord-blue.svg)](https://discord.gg/qGpsxSA)\n[![testgen](https://github.com/ethereum/consensus-specs/actions/workflows/generate_vectors.yml/badge.svg?branch=dev&event=schedule)](https://github.com/ethereum/consensus-specs/actions/workflows/generate_vectors.yml)\n\nThis repository hosts the current Ethereum\n[proof-of-stake](https://ethereum.org/en/developers/docs/consensus-mechanisms/pos/)\nspecifications. Discussions about design rationale and proposed changes can be\nbrought up and discussed as issues. Solidified, agreed-upon changes to the\nspecifications can be made through pull requests.\n\n## Specifications\n\nCore specifications for Ethereum proof-of-stake clients can be found in\n[specs](specs). These are divided into features. Features are researched and\ndeveloped in parallel, and then consolidated into sequential upgrades when\nread

In [55]:
# import os
# from mistralai import Mistral

# # api_key = os.environ["MISTRAL_API_KEY"]
# api_key = MISTRAL_API
# model = "mistral-large-latest"

# client = Mistral(api_key=api_key)

# stream_response = client.chat.stream(
#     model = model,
#     messages = [
#         {
#             "role": "user",
#             "content": "What is the best French cheese?",
#         },
#     ]
# )

# for chunk in stream_response:
#     print(chunk.data.choices[0].delta.content)

In [56]:
# import os
# from mistralai import Mistral

# # api_key = os.environ["MISTRAL_API_KEY"]
# model = "mistral-large-latest"

# client = Mistral(api_key=api_key)
# messages = [
#     {
#         "role": "user",
#         "content": "What is the best French meal? Return the name and the ingredients in short JSON object.",
#     }
# ]
# chat_response = client.chat.complete(
#       model = model,
#       messages = messages,
#       response_format = {
#           "type": "json_object",
#       }
# )

# print(chat_response.choices[0].message.content)



In [57]:
def analyze_project(readme_content, client):
    """Analyzes project documentation using freely avilable open-source models"""
    prompt = """Rate this open source project on each dimension (1-5). Use all available vibes and context to make judgment calls. Being decisive is better than being neutral. Use 3 only when there's truly no signal about a dimension.

Return JSON format:
{
 "technical_complexity": N,  // System complexity 
 "web3_focus": N,           // Blockchain/crypto focus
 "developer_tool": N,       // Developer infrastructure
 "project_maturity": N,     // Production readiness
 "community_size": N,       // Community scale
 "enterprise_ready": N,     // Production/enterprise grade
 "community_engagement": N, // Activity level
 "documentation": N,       // Docs quality
 "code_quality": N,        // Testing/standards
 "status": N,             // Project prestige
 "corporate": N,          // Corporate vs community
 "security": N,           // Security emphasis
 "innovation": N,         // Novel approaches
 "performance": N,        // Speed/efficiency focus
 "modularity": N,         // Reusability
 "accessibility": N,      // New contributor friendly
 "key_features": ["1-3 key features"]
}

Scale:
1 = Very Low
2 = Low
3 = Medium/No Signal
4 = High
5 = Very High

Project Documentation:
""" + readme_content

    response = client.chat.complete(
        messages=[{"role": "user", "content": prompt}],
        # model="gpt-4o",
        model = model,
        temperature=0.7
    )
    
    return response.choices[0].message.content


In [58]:
# import json

# result = analyze_project(readme_content[0], client)

# # Try to pretty-print JSON if possible, else print as-is
# try:
#     # Remove code block markers if present
#     cleaned = result.strip()
#     if cleaned.startswith("```json"):
#         cleaned = cleaned[7:]
#     if cleaned.startswith("```"):
#         cleaned = cleaned[3:]
#     if cleaned.endswith("```"):
#         cleaned = cleaned[:-3]
#     parsed = json.loads(cleaned)
#     print(json.dumps(parsed, indent=2, ensure_ascii=False))
# except Exception:
#     print(result)

In [59]:
prompt_for_juror = """ 

You are a expert juror in the field of open source project evaluation, you need to evaluate the project based on the following criteria

"""

In [60]:
# messages = [
#     {
#         "role": "user",
#         "content": prompt
#     }
# ]

# chat_response = client.chat.complete(
#       model = model,
#       messages = messages,
#       response_format = {
#           "type": "json_object",
#       }
# )

# print(chat_response.choices[0].message.content)

In [61]:
def get_readme(repo_url, headers=None):
    """Fetches README content from GitHub API"""
    # Extract owner and repo from github URL
    _, _, _, owner, repo = repo_url.rstrip('/').split('/')
    
    # List of possible README filenames
    readme_names = ['README.md', 'Readme.md', 'readme.md', 'README.markdown']
    
    headers = headers or {'Accept': 'application/vnd.github.v3+json'}
    
    for name in readme_names:
        try:
            url = f'https://api.github.com/repos/{owner}/{repo}/contents/{name}'
            response = requests.get(url, headers=headers)
            
            if response.status_code == 200:
                content = base64.b64decode(response.json()['content']).decode('utf-8')
                return content
                
        except Exception as e:
            continue
    
    return None

In [62]:
repo_with_features.columns

Index(['owner', 'name', 'dependencies', 'description', 'latestCommit',
       'githubLink', 'websiteLink', 'reputation', 'activity', 'popularity',
       'languages', 'commitCount', 'totalStars', 'recentContributors',
       'githubLink_x', 'total_avg_followers_by_contributor',
       'total_avg_commit_by_contributor', 'total_avg_activity_by_contributor',
       'total_avg_stars_by_contributor',
       'total_avg_reputation_by_contributor'],
      dtype='object')

In [63]:
train.shape

(126, 8)

In [64]:
repo_with_features['description'][5]

'Optional static typing for Python'

In [65]:
repo_with_features

owner             name dependencies  \
0                    oven-sh              bun    no status   
1                   zloirock          core-js    no status   
2                paradigmxyz             reth    no status   
3                  microsoft       playwright    no status   
4                     trezor  trezor-firmware    no status   
...                      ...              ...          ...   
4422                  coreos       go-systemd    no status   
4423             cockroachdb           errors    no status   
4424              manifoldco         promptui    no status   
4425             cockroachdb          logtags    no status   
4426  census-instrumentation    opencensus-go    no status   

                                            description latestCommit  \
0     Incredibly fast JavaScript runtime, bundler, t...     3/1/2025   
1                                      Standard Library     3/1/2025   
2     Modular, contributor-friendly and blazing-fast...     3/1/2025   
3     Playwright is a framework for Web Testing and ...     3/1/2025   
4                       :lock: Trezor Firmware Monorepo    2/28/2025   
...                                                 ...          ...   
4422  Go bindings to systemd socket activation, jour...   no commits   
4423  Go error library with error portability over t...     6/3/2024   
4424   Interactive prompt for command-line applications   no commits   
4425              key/value annotations for Go contexts   12/15/2024   
4426  A stats collection and distributed tracing fra...   no commits   

                                             githubLink  \
0                        https://github.com/oven-sh/bun   
1                   https://github.com/zloirock/core-js   
2                   https://github.com/paradigmxyz/reth   
3               https://github.com/microsoft/playwright   
4             https://github.com/trezor/trezor-firmware   
...                                                 ...   
4422               https://github.com/coreos/go-systemd   
4423              https://github.com/cockroachdb/errors   
4424             https://github.com/manifoldco/promptui   
4425             https://github.com/cockroachdb/logtags   
4426  https://github.com/census-instrumentation/open...   

                  websiteLink  reputation  activity  popularity  \
0              https://bun.sh          10        75          10   
1                         NaN          12        74          10   
2                         NaN           7        74          10   
3      https://playwright.dev          10        73          10   
4           https://trezor.io           6        73          10   
...                       ...         ...       ...         ...   
4422                      NaN           6         0          10   
4423                      NaN           8         0          10   
4424  https://www.manifold.co           5         0          10   
4425                      NaN           8         0           4   
4426     http://opencensus.io           6         0          10   

                                              languages  commitCount  \
0            Zig, JavaScript, TypeScript, CSS, Makefile        352.0   
1                          JavaScript, HTML, TypeScript        359.0   
2     Rust, Dockerfile, Makefile, JavaScript, Handle...        350.0   
3              JavaScript, Shell, TypeScript, HTML, CSS        299.0   
4                  Python, C, Makefile, Nix, Dockerfile        293.0   
...                                                 ...          ...   
4422                                          Go, Shell          0.0   
4423                                                 Go          3.0   
4424                                       Go, Makefile          0.0   
4425                                                 Go          2.0   
4426                          Go, Shell, HTML, Makefile          0.0   

      totalStars                  

In [ ]:
repo_feature_metadata=repo_with_features.drop(columns=['owner','name', 'dependencies', 'websiteLink', 'githubLink_x'])

repo_feature_metadata

description latestCommit  \
0     Incredibly fast JavaScript runtime, bundler, t...     3/1/2025   
1                                      Standard Library     3/1/2025   
2     Modular, contributor-friendly and blazing-fast...     3/1/2025   
3     Playwright is a framework for Web Testing and ...     3/1/2025   
4                       :lock: Trezor Firmware Monorepo    2/28/2025   
...                                                 ...          ...   
4422  Go bindings to systemd socket activation, jour...   no commits   
4423  Go error library with error portability over t...     6/3/2024   
4424   Interactive prompt for command-line applications   no commits   
4425              key/value annotations for Go contexts   12/15/2024   
4426  A stats collection and distributed tracing fra...   no commits   

                                             githubLink  reputation  activity  \
0                        https://github.com/oven-sh/bun          10        75   
1                   https://github.com/zloirock/core-js          12        74   
2                   https://github.com/paradigmxyz/reth           7        74   
3               https://github.com/microsoft/playwright          10        73   
4             https://github.com/trezor/trezor-firmware           6        73   
...                                                 ...         ...       ...   
4422               https://github.com/coreos/go-systemd           6         0   
4423              https://github.com/cockroachdb/errors           8         0   
4424             https://github.com/manifoldco/promptui           5         0   
4425             https://github.com/cockroachdb/logtags           8         0   
4426  https://github.com/census-instrumentation/open...           6         0   

      popularity                                          languages  \
0             10         Zig, JavaScript, TypeScript, CSS, Makefile   
1             10                       JavaScript, HTML, TypeScript   
2             10  Rust, Dockerfile, Makefile, JavaScript, Handle...   
3             10           JavaScript, Shell, TypeScript, HTML, CSS   
4             10               Python, C, Makefile, Nix, Dockerfile   
...          ...                                                ...   
4422          10                                          Go, Shell   
4423          10                                                 Go   
4424          10                                       Go, Makefile   
4425           4                                                 Go   
4426          10                          Go, Shell, HTML, Makefile   

      commitCount  totalStars  \
0           352.0     76508.0   
1           359.0     24860.0   
2           350.0      2722.0   
3           299.0     69737.0   
4           293.0      1437.0   
...           ...         ...   
4422          0.0      2474.0   
4423          3.0      2168.0   
4424          0.0      6166.0   
4425          2.0        39.0   
4426          0.0      2049.0   

                                     recentContributors  \
0     jprinaldi, luavixen, robertshuford, bjornthepr...   
1     slowcheetah, zloirock, szepeviktor, wchenonly,...   
2     zakisaad, yaziciahmet, thegaram, abnerzheng, m...   
3     aslushnikov, romarionijim, matthijskok, przemy...   
4     gabrielkerekes, mgiannopoulos24, obrusvit, sha...   
...                                                 ...   
4422                                                NaN   
4423                                         dhartunian   
4424                                                NaN   
4425                                        raduberinde   
4426                                                NaN   

      total_avg_followers_by_contributor  total_avg_commit_by_contributor  \
0                             167.000000                       243.869955   
1                            1248.833333                       262.333333   
2               

In [72]:
# No, but why haven't you vectorized it yet ??

from sklearn.feature_extraction.text import TfidfVectorizer
import numpy as np

# Fill NaNs in 'description' with empty strings
descriptions = repo_feature_metadata['description'].fillna('')

# Vectorize the descriptions using TF-IDF
vectorizer = TfidfVectorizer()
vectors = vectorizer.fit_transform(descriptions)

# Add the vectorized descriptions as a new column
repo_feature_metadata['description_vector'] = list(vectors.toarray())


In [79]:
descriptions[0]

_ = vectorizer.fit_transform([descriptions[0]])
_.toarray()

array([[0.2773501, 0.2773501, 0.2773501, 0.2773501, 0.2773501, 0.2773501,
        0.2773501, 0.2773501, 0.2773501, 0.2773501, 0.2773501, 0.2773501,
        0.2773501]])

In [73]:
repo_feature_metadata['description_vector']

0       [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
1       [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
2       [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
3       [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
4       [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
                              ...                        
4422    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
4423    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
4424    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
4425    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
4426    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
Name: description_vector, Length: 4427, dtype: object